elements/lisk-codec/test/string.spec.ts

```diff
@@ -49,5 +50,82 @@ describe('string', () => {
 		});
+
+		describe('invalid utf-8 code', () => {
+			const testCases = [
+				// Valid sequences
+				{ input: [0x00], expected: true }, // 1-byte sequence
+				{ input: [0x7f], expected: true }, // 1-byte sequence (max value)
+				{ input: [0xc2, 0x80], expected: true }, // 2-byte sequence
+				{ input: [0xdf, 0xbf], expected: true }, // 2-byte sequence (max value)
+				{ input: [0xe0, 0xa0, 0x80], expected: true }, // 3-byte sequence
+				{ input: [0xef, 0xbf, 0xbf], expected: true }, // 3-byte sequence (max value)
+				{ input: [0xf0, 0x90, 0x80, 0x80], expected: true }, // 4-byte sequence U+10FFFF
+				{ input: [0xf4, 0x8f, 0xbf, 0xbf], expected: true }, // 4-byte sequence (max value)
+
+				// Invalid sequences
+				{ input: [0x80], expected: false }, // standalone continuation byte
+				{ input: [0xbf], expected: false }, // standalone continuation byte (max value)
+				{ input: [0xc0, 0x80], expected: false }, // overlong encoding (2-byte sequence)
+				{ input: [0xe0, 0x80, 0x80], expected: false }, // overlong encoding (3-byte sequence)
+				{ input: [0xf0, 0x80, 0x80, 0x80], expected: false }, // overlong encoding (4-byte sequence)
+				{ input: [0xc1, 0xbf], expected: false }, // overlong encoding (2-byte sequence, max value)
+				{ input: [0xe0, 0x9f, 0xbf], expected: false }, // overlong encoding (3-byte sequence, max value)
+				{ input: [0xf0, 0x8f, 0xbf, 0xbf], expected: false }, // overlong encoding (4-byte sequence, max value)
+				{ input: [0xc2], expected: false }, // missing continuation byte (2-byte sequence)
+				{ input: [0xe0, 0xa0], expected: false }, // missing continuation byte (3-byte sequence)
+				{ input: [0xf0, 0x90, 0x80], expected: false }, // missing continuation byte (4-byte sequence)
+				{ input: [0xc2, 0xc0], expected: false }, // invalid continuation byte (2-byte sequence)
+				{ input: [0xe0, 0xa0, 0xc0], expected: false }, // invalid continuation byte (3-byte sequence)
+				{ input: [0xf0, 0x90, 0x80, 0xc0], expected: false }, // invalid continuation byte (4-byte sequence)
+			];
+			it.each(testCases)('given $input should return $expected', ({ input, expected }) => {
+				const buffer = Buffer.from(input);
+				const inputBytes = writeBytes(buffer);
+				if (expected) {
+					expect(() => readString(inputBytes, 0)).not.toThrow();
+				} else {
+					expect(() => readString(inputBytes, 0)).toThrow(
+						'The encoded data was not valid for encoding utf-8',
+					);
+				}
+			});
+		});
+
+		describe('unnormalized utf-8 code', () => {
+			const testCases = [
+				// Valid sequences (NFC)
+				{ input: 'Amélie'.normalize('NFC'), expected: true }, // "Amélie" in composed form (NFC)
+				{ input: 'Ame\u0301lie', expected: false }, // "Amélie" in decomposed form (NFD)
+
+				// Invalid sequences (NFD)
+				{ input: 'Straße'.normalize('NFC'), expected: true }, // "Straße" in composed form (NFC)
+				{ input: 'Strasse\u0308', expected: false }, // "Straße" in decomposed form (NFD)
```

Comment:
Is this example wrong? `e\u0308` puts two dots over the `e`. I guess you wanted to have two encodings for `ß`?

Reply:
Actually `ß` is special case, which is the same for NFC and NFD. Updated the test case